File: libs/cua-driver/examples/jev-use/python/visual_adapter.py
from __future__ import annotations

from typing import Any

from core import Candidate


def _valid_box(box: dict[str, Any], width: int, height: int) -> bool:
    values = [box.get(key) for key in ("x", "y", "width", "height")]
    if not all(isinstance(value, int) for value in values):
        return False
    x, y, box_width, box_height = values
    return x >= 0 and y >= 0 and box_width > 0 and box_height > 0 and x + box_width <= width and y + box_height <= height


def _overlaps(left: dict[str, int], right: dict[str, int]) -> bool:
    return (
        left["x"] < right["x"] + right["width"]
        and right["x"] < left["x"] + left["width"]
        and left["y"] < right["y"] + right["height"]
        and right["y"] < left["y"] + left["height"]
    )


def build_visual_candidates(
    observation: dict[str, Any], response: dict[str, Any], *, minimum_confidence: float = 0.8
) -> list[Candidate]:
    fallback = [
        Candidate("reobserve", "Capture fresh evidence before proposing another action.", None, {}),
        Candidate("abstain", "Stop without acting when the evidence is unsafe or ambiguous.", None, {}),
    ]
    capture = response.get("capture", {})
    if response.get("schema") != "cua.visual_regions_v1" or any(
        capture.get(key) != observation.get(key)
        for key in ("source", "snapshot_id", "image_width", "image_height")
    ):
        return fallback

    width = observation.get("image_width")
    height = observation.get("image_height")
    source = observation.get("source", {})
    if (
        not isinstance(width, int)
        or not isinstance(height, int)
        or source.get("kind") != "window"
        or not isinstance(source.get("pid"), int)
        or not isinstance(source.get("window_id"), int)
    ):
        return fallback

    semantic = observation.get("semantic_elements", [])
    matches: list[dict[str, Any]] = []
    for region in response.get("regions", []):
        box = region.get("bounds", {})
        center = region.get("center", {})
        if (
            region.get("interactive") is not True
            or not isinstance(region.get("confidence"), (int, float))
            or region["confidence"] < minimum_confidence
            or not _valid_box(box, width, height)
            or not isinstance(center.get("x"), int)
            or not isinstance(center.get("y"), int)
            or not (box["x"] <= center["x"] < box["x"] + box["width"])
            or not (box["y"] <= center["y"] < box["y"] + box["height"])
        ):
            continue
        aligned = [
            element
            for element in semantic
            if str(element.get("name", "")).casefold() == str(region.get("text", "")).casefold()
            and _valid_box(element.get("bounds", {}), width, height)
            and _overlaps(box, element["bounds"])
        ]
        if len(aligned) == 1:
            matches.append(region)

    if len(matches) != 1:
        return fallback
    region = matches[0]
    action = Candidate(
        "click-visual-save",
        "Click the fresh visual region labelled Save.",
        "click",
        {
            "pid": source["pid"],
            "window_id": source["window_id"],
            "x": region["center"]["x"],
            "y": region["center"]["y"],
            "delivery_mode": "background",
        },
    )
    return [action, *fallback]

File: libs/cua-driver/examples/jev-use/typescript/visual-adapter.ts
import type { Candidate } from './core.js';

type Bounds = { x: number; y: number; width: number; height: number };
type Observation = {
  source: { kind: string; pid?: number; window_id?: number };
  snapshot_id: string;
  image_width: number;
  image_height: number;
  semantic_elements?: { name?: string; bounds?: Bounds }[];
};

function validBox(value: unknown, width: number, height: number): value is Bounds {
  if (!value || typeof value !== 'object') return false;
  const box = value as Record<string, unknown>;
  const values = ['x', 'y', 'width', 'height'].map((key) => box[key]);
  if (!values.every(Number.isInteger)) return false;
  const [x, y, boxWidth, boxHeight] = values as number[];
  return x >= 0 && y >= 0 && boxWidth > 0 && boxHeight > 0 && x + boxWidth <= width && y + boxHeight <= height;
}

function overlaps(left: Bounds, right: Bounds): boolean {
  return left.x < right.x + right.width && right.x < left.x + left.width && left.y < right.y + right.height && right.y < left.y + left.height;
}

export function buildVisualCandidates(
  observation: Observation,
  response: Record<string, any>,
  minimumConfidence = 0.8
): Candidate[] {
  const fallback: Candidate[] = [
    { id: 'reobserve', description: 'Capture fresh evidence before proposing another action.', tool: null, arguments: {} },
    { id: 'abstain', description: 'Stop without acting when the evidence is unsafe or ambiguous.', tool: null, arguments: {} },
  ];
  const capture = response.capture ?? {};
  if (
    response.schema !== 'cua.visual_regions_v1' ||
    JSON.stringify(capture.source) !== JSON.stringify(observation.source) ||
    capture.snapshot_id !== observation.snapshot_id ||
    capture.image_width !== observation.image_width ||
    capture.image_height !== observation.image_height ||
    observation.source.kind !== 'window' ||
    !Number.isInteger(observation.source.pid) ||
    !Number.isInteger(observation.source.window_id)
  ) return fallback;

  const matches = (response.regions ?? []).filter((region: Record<string, any>) => {
    const box = region.bounds;
    const center = region.center ?? {};
    if (
      region.interactive !== true ||
      typeof region.confidence !== 'number' ||
      region.confidence < minimumConfidence ||
      !validBox(box, observation.image_width, observation.image_height) ||
      !Number.isInteger(center.x) ||
      !Number.isInteger(center.y) ||
      center.x < box.x || center.x >= box.x + box.width ||
      center.y < box.y || center.y >= box.y + box.height
    ) return false;
    const aligned = (observation.semantic_elements ?? []).filter(
      (element) =>
        element.name?.toLocaleLowerCase() === String(region.text ?? '').toLocaleLowerCase() &&
        validBox(element.bounds, observation.image_width, observation.image_height) &&
        overlaps(box, element.bounds)
    );
    return aligned.length === 1;
  });
  if (matches.length !== 1) return fallback;
  const region = matches[0];
  return [
    {
      id: 'click-visual-save',
      description: 'Click the fresh visual region labelled Save.',
      tool: 'click',
      arguments: {
        pid: observation.source.pid,
        window_id: observation.source.window_id,
        x: region.center.x,
        y: region.center.y,
        delivery_mode: 'background',
      },
    },
    ...fallback,
  ];
}
